File: src/server/Hoard.Core/Application/Portfolios/PortfolioPerformanceDto.cs
namespace Hoard.Core.Application.Portfolios;

public class PortfolioPerformanceDto
{
    // Snapshot values
    public decimal Value { get; init; }
    public decimal CashValue { get; init; }
    public decimal CashPercentage { get; set; }
    public decimal PreviousValue { get; init; }
    public decimal ValueChange { get; init; }

    // Gains
    public decimal UnrealisedGain { get; init; }
    public decimal RealisedGain { get; init; }
    public decimal Income { get; init; }

    // Absolute value changes
    public decimal? ValueChange1Y { get; init; }

    // Returns (nullable = not enough data yet)
    public decimal? Return1D { get; init; }
    public decimal? Return1W { get; init; }
    public decimal? Return1M { get; init; }
    public decimal? Return3M { get; init; }
    public decimal? Return6M { get; init; }
    public decimal? Return1Y { get; init; }
    public decimal? Return3Y { get; init; }
    public decimal? Return5Y { get; init; }
    public decimal? Return10Y { get; init; }
    public decimal? ReturnYtd { get; init; }
    public decimal? ReturnAllTime { get; init; }
    public decimal? AnnualisedReturn { get; init; }

    public DateTime UpdatedUtc { get; init; }
}

File: src/client/hoard-ui/src/api/dtos/PortfolioPerformanceDto.ts
export interface PortfolioPerformanceDto {
  // Snapshot values
  value: number;
  cashValue: number;
  cashPercentage: number;
  previousValue: number;
  valueChange: number;

  // Gains
  unrealisedGain: number;
  realisedGain: number;
  income: number;

  // Absolute value changes
  valueChange1Y: number | null;

  // Returns (nullable = not enough data yet)
  return1D: number | null;
  return1W: number | null;
  return1M: number | null;
  return3M: number | null;
  return6M: number | null;
  return1Y: number | null;
  return3Y: number | null;
  return5Y: number | null;
  return10Y: number | null;
  returnYtd: number | null;
  returnAllTime: number | null;
  annualisedReturn: number | null;

  updatedUtc: string; // ISO string
}
